interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,877 @@
+	rdb.perUser = make(map[uint32]*userDB)
+	for _, rule := range origRules {
+		_, exists := modifiedUserRuleIDs[rule.User]
+		if !exists {
+			modifiedUserRuleIDs[rule.User] = make(map[prompting.IDType]map[string]string)
+		}
+		if notifyEveryRule {
+			modifiedUserRuleIDs[rule.User][rule.ID] = nil
+		}
+		if err := rule.Lifespan.ValidateExpiration(rule.Expiration, currTime); err != nil || rule.Constraints.ValidateForInterface(rule.Interface) != nil {
+			// Invalid rule, discard it
+			needToSave = true
+			data := map[string]string{"removed": "invalid"}
+			if errors.Is(err, prompting.ErrExpirationInThePast) {
+				// Not actually invalid, just expired
+				data["removed"] = "expired"
+			}
+			modifiedUserRuleIDs[rule.User][rule.ID] = data
+			continue
+		}
+		for {
+			err, conflicts, conflictingPermission := rdb.addRuleToTree(rule)
+			if err == nil {
+				break
+			}
+			// Err must be ErrPathPatternConflict.
+			// Prioritize newer rules by pruning permission from old rule until
+			// no conflicts remain.
+			// XXX: this results in the permission being dropped for all other
+			// variants of the older rule.
+			// TODO: split older rule into two rules, preserving all except the
+			// directly conflicting variant/permission combination.
+			for _, conflict := range conflicts {
+				conflictingID := conflict.ConflictingID
+				conflictingRule, _ := rdb.ruleWithID(conflictingID) // must exist
+				if rule.Timestamp.After(conflictingRule.Timestamp) {
+					rdb.removeRulePermissionFromTree(conflictingRule, conflictingPermission) // must return nil
+					var data map[string]string
+					if conflictingRule.removePermission(conflictingPermission) == prompting.ErrPermissionsListEmpty {
+						rdb.removeRuleWithID(conflictingID)
+						data = map[string]string{"removed": "conflict"}
+					}
+					modifiedUserRuleIDs[conflictingRule.User][conflictingID] = data
+				} else {
+					rule.removePermission(conflictingPermission) // ignore error
+					var data map[string]string
+					modifiedUserRuleIDs[rule.User][rule.ID] = data
+				}
+				needToSave = true
+			}
+		}
+		if len(rule.Constraints.Permissions) > 0 {
+			rdb.addRule(rule)
+		} else {
+			// TODO: record status of the rule ("removed") in modifiedUserRuleIDs
+			needToSave = true
+			data := map[string]string{"removed": "conflict"}
+			modifiedUserRuleIDs[rule.User][rule.ID] = data
+		}
+	}
+}
+
+// PopulateNewRule creates a new Rule with the given contents.
+//
+// Users of requestrules should probably autofill rules from JSON and never call
+// this function directly.
+//
+// Constructs a new rule with the given parameters as values, with the
+// exception of duration. Uses the given duration, in addition to the current
+// time, to compute the expiration time for the rule, and stores that as part
+// of the rule which is returned. If any of the given parameters are invalid,
+// returns a corresponding error.
+//
+// TODO: unexport (probably, avoid confusion with AddRule)
```

Comment:
this definitely should not be exported as is not taking the lock

Reply:
Holding the lock isn't technically required here, since it's just getting the next ID, doing some validation, and populating a struct, and getting the next ID is atomic. Better to unexport though, nonetheless.